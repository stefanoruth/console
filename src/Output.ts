import rl from 'readline'

const colorList = {
	reset: '\x1b[0m',
	cyan: '\x1b[36m',
	yellow: '\x1b[33m',
}

type Colors = keyof typeof colorList

export class Output {
	/**
	 * Write a line to the console.
	 */
	line(message: string, newLine: boolean = false, color: Colors = 'reset'): void {
		process.stdout.write(color + '%s' + colorList.reset, message)

		if (newLine) {
			this.newLine()
		}
	}

	/**
	 * Display a table on the console.
	 */
	table(...args: any[]): void {
		console.table(args)
	}

	/**
	 * Display a set of new lines.
	 */
	newLine(count: number = 1) {
		process.stdout.write('\n'.repeat(count))
	}

	/**
	 * Ask the user for a question.
	 */
	ask(question: string): Promise<string> {
		return new Promise(resolve => {
			const r = rl.createInterface({
				input: process.stdin,
				output: process.stdout,
			})

			r.question(question + '\n', (answer: string) => {
				r.close()
				resolve(answer)
			})
		})
	}

	/**
	 * Ask the user if they really wanna do it.
	 */
	confirm(question: string, defaultValue: boolean = true): Promise<boolean> {
		return new Promise(resolve => {
			const r = rl.createInterface({
				input: process.stdin,
				output: process.stdout,
			})

			r.question(question + '\n', (answer: string) => {
				r.close()
				resolve(answer === 'Y')
			})
		})
	}
}
