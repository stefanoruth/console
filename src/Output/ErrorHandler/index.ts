export * from './FilePreview'
export * from './StackTrace'
export * from './ErrorHandler'
