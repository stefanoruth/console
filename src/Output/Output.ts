import rl from 'readline'
import { Color, CliColor } from './CliColor'
import { ProgressBar } from './ProgressBar'
import { Table } from './Table'
import { Question } from './Question/Question'
import { Writer } from './Writer'

export class Output {
	constructor(public writer: Writer = new Writer(), protected color: CliColor = new CliColor()) {}

	/**
	 * Write a line to the console.
	 */
	line(message: string, newLine: boolean = true, color?: Color) {
		this.writer.write(this.color.apply(message, { text: color }), newLine)
	}

	/**
	 * Formats a command title.
	 */
	title(message: string) {
		//
	}

	/**
	 * Formats a section title.
	 */
	section(message: string) {
		//
	}

	/**
	 * Formats a list.
	 */
	listing(elements: any[]) {
		//
	}

	/**
	 * Formats a command comment.
	 */
	comment(message: string | string[]) {
		this.writer.block(message, undefined, undefined, '<fg=default;bg=default> // </>', false, false)
	}

	/**
	 * Formats a success result bar.
	 */
	success(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'green' }))
	}

	/**
	 * Formats an error result bar.
	 */
	error(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'white', bg: 'red' }))
		// this.block(message, 'ERROR', 'fg=white;bg=red', ' ', true)
	}

	/**
	 * Formats an warning result bar.
	 */
	warning(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'black', bg: 'yellow' }))
		// this.block(message, 'WARNING', 'fg=black;bg=yellow', ' ', true)
	}

	/**
	 * Formats a note admonition.
	 */
	note(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'yellow' }))
		// this.block(message, 'NOTE', 'fg=yellow', ' ! ')
	}

	/**
	 * Formats a caution admonition.
	 */
	caution(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'white', bg: 'red' }))
		// this.block(message, 'CAUTION', 'fg=white;bg=red', ' ! ', true)
	}

	/**
	 * Display a table on the console.
	 */
	table(rows: object[], columns?: object) {
		const table = new Table(this)
		table.setHeaders(columns)
		table.setRows(rows)

		table.render()
		this.newLine()
	}

	/**
	 * Display a set of new lines.
	 */
	newLine(count: number = 1) {
		this.writer.newLine(count)
	}

	/**
	 * Ask the user for a question.
	 */
	ask(question: string) {
		return new Question().ask(question)
	}

	askQuestion(question: Question) {
		//
	}

	/**
	 * Ask the user if they really wanna do it.
	 */
	confirm(question: string, defaultValue: boolean = true): Promise<boolean> {
		return new Promise(resolve => {
			const r = rl.createInterface({
				input: process.stdin,
				output: process.stdout,
			})

			r.question(question + '\n', (answer: string) => {
				r.close()
				resolve(answer.toUpperCase() === 'Y')
			})
		})
	}

	/**
	 * Start a new progress bar
	 */
	progressBar(max: number = 0): ProgressBar {
		const progressBar = new ProgressBar(this, max)

		progressBar.start()

		return progressBar
	}
}
