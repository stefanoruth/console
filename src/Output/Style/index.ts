export * from './TextStyle'
export * from './Color'
export * from './NullColor'
