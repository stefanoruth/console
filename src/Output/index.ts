export * from './ErrorHandler'
export * from './ProgressBar'
export * from './Question'
export * from './Style'
export * from './Table'
export * from './Descriptor'
export * from './Output'
export * from './OutputFormatter'
export * from './Terminal'
export * from './Verbosity'
export * from './Writer'
