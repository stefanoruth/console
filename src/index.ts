export { Application } from './Application'
export { Input, Option, Signature, Argument } from './Input'
export { Output, Terminal, ProgressBar, Table } from './Output'
export { Command } from './Commands'
export * from './Events/EventCollection'
export * from './Exceptions'
