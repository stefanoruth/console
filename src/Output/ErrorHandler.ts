import { Output } from './Output'
import { CommandNotFoundException } from '../Exceptions'
import { Registry } from '../Commands/Registry'
import { Command } from '../Commands'
import { FilePreview } from './Error/FilePreview'
import { Stack } from './Error/Stack'

export class ErrorHandler {
	constructor(
		protected output: Output,
		protected filePreview: FilePreview = new FilePreview(output),
		protected stackRender: Stack = new Stack()
	) {}

	render(e: Error) {
		// console.log(e)

		if (e instanceof CommandNotFoundException) {
			return this.unknownCommand(e)
		}

		const style = this.output.getStyle()

		const error: string[] = []

		error.push(`\n  ${style.error(` ${e.name} `)} :  ${style.note(e.message)}\n\n`)

		error.push(...this.getErrorStack(e))

		this.output.raw(error)
	}

	runningCommand(e: Error, c: Command, applicationName: string) {
		this.output.newLine()
		this.output.info(c.getSignature().getSynopsis() + applicationName)

		this.render(e)

		this.output.newLine()
	}

	unknownCommand(e: CommandNotFoundException) {
		this.output.error(e.message)
	}

	suggestOtherCommands() {
		//
	}

	protected formatFile(file: string) {
		return file
	}

	protected getErrorStack(error: Error) {
		const style = this.output.getStyle()
		const render: string[] = []

		const tracing = this.stackRender.build(error)

		const entry = tracing.shift()!

		render.push(`  at ${style.success(`${this.formatFile(entry.file)}:${entry.line}:${entry.column}`)}`)
		render.push(...this.filePreview.view(entry))
		render.push('')

		render.push(`  ${style.info('Exception trace:')}\n`)

		tracing.forEach((trace, index) => {
			render.push(`  ${style.format(String(index), { text: 'blue' })}   ${style.info(this.formatFile(trace.method))}`)
			if (trace.file) {
				render.push(`${' '.repeat(6)}${style.success(`${this.formatFile(trace.file)}:${trace.line}:${trace.column}`)}`)
			}
			render.push('')
		})

		return render.map(item => item + '\n')
	}
}
