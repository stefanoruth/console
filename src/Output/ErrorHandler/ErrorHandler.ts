import { isRunningTestMode } from '../../helpers'
import { Output } from '../Output'
import { FilePreview } from './FilePreview'
import { StackTrace } from './StackTrace'

export interface ErrorHandlerContract {
	report: (error: Error) => void
}

export class ErrorHandler {
	constructor(protected output: Output) {}

	report(e: Error): string {
		if (isRunningTestMode()) {
			console.log('Running in test with error')
			// throw e
		}

		const preview = new FilePreview(this.output)
		const stackTrace = new StackTrace().render(e)
		const style = this.output.getStyle()

		const render: string[] = []

		render.push(`\n  ${style.error(` ${e.name} `)} :  ${style.note(e.message)}\n`)

		const entry = stackTrace.shift()!

		render.push(`  at ${style.success(`${entry.file}:${entry.line}:${entry.column}`)}`)
		render.push(...preview.render({ path: entry.file, line: entry.line }))
		render.push('')

		render.push(`  ${style.info('Exception trace:')}\n`)

		stackTrace.forEach((trace, index) => {
			render.push(`  ${style.format(index + '', { text: 'blue' })}   ${style.info(trace.method || '')}`)
			if (trace.file) {
				render.push(`${' '.repeat(6)}${style.success(`${trace.file}:${trace.line}:${trace.column}`)}`)
			}
			render.push('')
		})

		const errorMsg = render.join('\n')

		this.output.raw(errorMsg)

		return errorMsg
	}
}
