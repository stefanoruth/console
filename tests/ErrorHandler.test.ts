import { ErrorHandler } from '../src/Output/ErrorHandler'
import { Output, Terminal } from '../src/Output'
import { Writer } from '../src/Output/Writer'
import { TestThrow, TestThrowCustom } from './__mocks__/ThrowError'
import { Mock } from 'ts-mockery'
import { Formatter } from '../src/Output/Style'
import { TestColor, NullColor } from './__mocks__/TestColor'

function getHandler(cb: (message: string) => void) {
	const t = Mock.all<Terminal>()
	const w = new (class extends Writer {
		write(m: string | string[]) {
			if (!(m instanceof Array)) {
				m = [m]
			}

			cb(m.join(''))
		}
	})(t)

	return new (class extends ErrorHandler {
		protected formatFile(file: string) {
			return file.replace(process.cwd(), '/project')
		}
	})(new Output(t, w, new Formatter(new NullColor())))
}

describe('ErrorHandler', () => {
	test('Render Basic Error', () => {
		try {
			TestThrow()
		} catch (error) {
			getHandler(m => {
				expect(m).toMatchSnapshot()
			}).render(error)
		}
	})
})
