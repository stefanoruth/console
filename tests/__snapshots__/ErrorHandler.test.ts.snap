// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler Render Basic Error 1`] = `
"
   Error  :  Foobar

  at /project/tests/__mocks__/ThrowError.ts:5:8

  Exception trace:

  0   Object.<anonymous>
      /project/tests/ErrorHandler.test.ts:31:4

  1   Object.asyncJestTest
      /project/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37

  2   /project/node_modules/jest-jasmine2/build/queueRunner.js:43:12

  3   new Promise (<anonymous>) mapper
      /project/node_modules/jest-jasmine2/build/queueRunner.js:26:19

  4   /project/node_modules/jest-jasmine2/build/queueRunner.js:73:41

  5   processTicksAndRejections
      internal/process/task_queues.js:89:5

"
`;
